File: files/problems/c/two-sum-iv-input-is-a-bst.c
void traverse(struct TreeNode *root, int *arr, int *index) {
    if (!root) return;
    traverse(root->left, arr, index);
    arr[(*index)++] = root->val;
    traverse(root->right, arr, index);
}

bool findTarget(struct TreeNode *root, int k) {
    int *arr = malloc(sizeof(int) * 10000), index = 0, sum = 0;
    traverse(root, arr, &index);
    int i = 0, j = index - 1;
    while (i < j) {
        sum = arr[i] + arr[j];
        if (sum > k) j--;
        else if (sum < k) i++;
        else break;
    }
    return i != j;
}

File: leetcode/problems/ts/two-sum-iv-input-is-a-bst.ts
// by @codeAbinash
// Time : O(n)
// Space : O(n)

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function findTarget(root: TreeNode | null, k: number): boolean {
  const set = new Set<number>();
  function traverse(node: TreeNode | null): boolean {
    if (!node) return false;
    if (set.has(k - node.val)) return true;
    set.add(node.val);
    return traverse(node.left) || traverse(node.right);
  }
  return traverse(root);
}
